File: samples/java-springboot/src/test/java/com/mongodb/reactor/client/sample/java/springboot/SampleServiceTest.java
package com.mongodb.reactor.client.sample.java.springboot;

import com.mongodb.reactivestreams.client.MongoClient;
import com.mongodb.reactivestreams.client.MongoCollection;
import org.bson.types.ObjectId;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.context.DynamicPropertySource;
import org.testcontainers.containers.MongoDBContainer;
import reactor.core.publisher.Mono;
import reactor.test.StepVerifier;

import java.util.ArrayList;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;

@SpringBootTest
class SampleServiceTest {

    private static final String DB_NAME = "test";

    private static final MongoDBContainer container = initContainer();

    @Autowired
    private SampleService service;

    @Autowired
    private MongoClient client;

    static MongoDBContainer initContainer() {
        MongoDBContainer container = new MongoDBContainer("mongo:4.4");
        container.start();
        return container;
    }

    @DynamicPropertySource
    static void databaseProperties(DynamicPropertyRegistry registry) {
        registry.add("spring.data.mongodb.port", container::getFirstMappedPort);
    }

    @BeforeEach
    void setup() {
        MongoCollection<User> collection = client.getDatabase(DB_NAME).getCollection("users", User.class);
        Mono.from(collection.insertMany(Stream.of(
                new User().setFirstName("John").setLastName("Doe").setEmail("john.doe@mail.com"),
                new User().setFirstName("Jane").setLastName("Doe").setEmail("jane.doe@mail.com")
        ).collect(Collectors.toList()))).block();
    }

    @AfterEach
    void cleanUp() {
        Mono.from(client.getDatabase(DB_NAME).getCollection("users").drop()).block();
    }

    @Test
    void listDatabasesNamesShouldReturnTestDbName() {
        StepVerifier.create(service.listDatabasesNames())
                    .expectSubscription()
                    .recordWith(ArrayList::new)
                    .thenConsumeWhile(s -> true)
                    .consumeRecordedWith(l -> {
                        assertThat(l.size()).isGreaterThan(1);
                        assertThat(l).contains(DB_NAME);
                    })
                    .verifyComplete();
    }

    @Test
    void listCollectionsShouldReturnUsersCollection() {
        StepVerifier.create(service.listCollections())
                    .expectSubscription()
                    .expectNext("users")
                    .verifyComplete();
    }

    @Test
    void findAllShouldReturnMultipleUsers() {
        StepVerifier.create(service.findAll())
                    .expectSubscription()
                    .recordWith(ArrayList::new)
                    .expectNextCount(2)
                    .consumeRecordedWith(l -> assertThat(l.stream().map(User::getFirstName)).contains("John", "Jane"))
                    .verifyComplete();
    }

    @Test
    void createShouldAddNewDocument() {
        User user = new User().setFirstName("Mark").setLastName("Doe").setEmail("mark.doe@gmail.com").setId(new ObjectId());
        StepVerifier.create(service.create(user))
                    .expectSubscription()
                    .consumeNextWith(u -> assertThat(u).usingRecursiveComparison().isEqualTo(user))
                    .verifyComplete();
    }
}

File: samples/kotlin-springboot/src/test/kotlin/com/mongodb/client/reactor/sample/kotlin/springboot/SampleServiceTest.kt
package com.mongodb.client.reactor.sample.kotlin.springboot

import com.mongodb.reactivestreams.client.MongoClient
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.DynamicPropertyRegistry
import org.springframework.test.context.DynamicPropertySource
import org.testcontainers.containers.MongoDBContainer
import reactor.kotlin.core.publisher.toMono
import reactor.kotlin.test.test

@SpringBootTest
internal class SampleServiceTest(@Autowired private val service: SampleService, @Autowired private val client: MongoClient) {

    companion object {

        private const val DB_NAME = "test"
        private val container = MongoDBContainer("mongo:4.4").apply { start() }

        @JvmStatic
        @DynamicPropertySource
        fun databaseProperties(registry: DynamicPropertyRegistry) {
            registry.add("spring.data.mongodb.port") { container.firstMappedPort }
        }
    }

    @BeforeEach
    fun setup() {
        val collection = client.getDatabase(DB_NAME).getCollection("users", User::class.java)
        collection.insertMany(
            listOf(
                User("john.doe@mail.com", "John", "Doe"),
                User("jane.doe@mail.com", "Jane", "Doe")
            )
        ).toMono().block()
    }

    @AfterEach
    fun cleanUp() {
        client.getDatabase(DB_NAME).getCollection("users").drop().toMono().block()
    }

    @Test
    fun `list databases names should return test db name`() {
        service.listDatabasesNames().test()
            .expectSubscription()
            .recordWith { ArrayList() }
            .thenConsumeWhile { true }
            .consumeRecordedWith {
                assertThat(it.size).isGreaterThan(1)
                assertThat(it).contains(DB_NAME)
            }
            .verifyComplete()
    }

    @Test
    fun listCollectionsShouldReturnUsersCollection() {
        service.listCollections().test()
            .expectSubscription()
            .expectNext("users")
            .verifyComplete()
    }

    @Test
    fun findAllShouldReturnMultipleUsers() {
        service.findAll().test()
            .expectSubscription()
            .recordWith { ArrayList() }
            .expectNextCount(2)
            .consumeRecordedWith { s -> assertThat(s.stream().map { it.firstName }).contains("John", "Jane") }
            .verifyComplete()
    }

    @Test
    fun createShouldAddNewDocument() {
        val user = User("mark.doe@mail.com", "Mark", "Doe")
        service.create(user).test()
            .expectSubscription()
            .consumeNextWith { assertThat(it).usingRecursiveComparison().isEqualTo(user) }
            .verifyComplete()
    }
}
